src/migrateDashboard.ts

```diff
@@ -56,7 +59,22 @@ export function migrateDashboard(
           dashboardLeafKey,
         ];
       } else {
-        content[dashboardLeafKey] = migrateWidget(widget.bookmark, servers);
+        const [migratedWidget, widgetMigrationError] = migrateWidget(
+          widget.bookmark,
+          servers,
+        );
+        content[dashboardLeafKey] = migratedWidget;
+        if (widgetMigrationError) {
+          if (!dashboardMigrationReport.pages) {
+            dashboardMigrationReport.pages = {};
+          }
+          if (!dashboardMigrationReport.pages[pageKey]) {
+            dashboardMigrationReport.pages[pageKey] = { warnings: [] };
+          }
+          dashboardMigrationReport.pages[pageKey].warnings.push(
```

Comment:
It seems misleading to push an error into an attribute named `warnings`. I would suggest to change the report type as follows:

```
interface FileErrorReport {
  // The id of each parent folder, from the root down to the direct parent.
  folderId: string[];
  // The name of each parent folder, from the root down to the direct parent.
  folderName: string[];
  // The name of the file on which the error occurred.
  name: string;
  // The thrown error.
  error: Error;
}

interface MigrationReport {
  dashboards: {
    [dashboardId]: 
      // If the error was thrown by `migrateDashboard` itself, not an underlying call to `migrateWidget`.
      // This should happen rarely.
      // In this case, the whole outdated dashboard is copied as is.
      | FileErrorReport
      // If the error was thrown by one or several underlying call(s) to `migrateWidget`.
      // This should happen more frequently.
      // In this case, the dashboard should be migrated succesfully, except for the widgets which threw errors.
      // These widgets should be copied as is into the new dashboard.
      | {
          pages: {
            [pageKey: string]:  {
              pageName: string;
              widgets: {
                [leafKey]: {
                  widgetName: string;
                  error: Error;
                }
              }
            }
          }
        }
  },
  widgets: {
    [widgetId: string]: FileErrorReport;
  },
  filters: {
    [filterId: string]: FileErrorReport;
  }
}
```

Reply:
Two remarks/questions about this:
- Do you intend to put `Error` objects in the file? I'm discovering that error properties are not enumerable and therefore do not appear once serialized to json, resulting in things like:
```
"widgets": {
            "0": {
              "widgetName": "dimensions/PNLs",
              "error": {
                "cubeName": "EquityDerivativesCube2",
                "availableCubeNames": [
                  "EquityDerivativesCube",
                  "EquityDerivativesCubeDist"
                ]
              }
            }
          }
```

To get a complete vision of the error we should use another structure to include the message and the stack.
- In the comment you say that an error with the dashboard migration results in the dashboard being copied "as-is". This is not done currently, probably because we do not expect errors outside of migrating widgets within a dashboard. It is not costly to do it but I don't really see the point.